[quant][fx] Enable input and output idx support dtype

terrychenism opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73816

Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34660121](https://our.internmc.facebook.com/intern/diff/D34660121)

terrychenism committed:
[quant][fx] Enable input and output idx support dtype

Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34660121).

terrychenism committed:
Update on "[quant][fx] Enable input and output idx support dtype"


Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34660121](https://our.internmc.facebook.com/intern/diff/D34660121)

[ghstack-poisoned]

vkuzo commented:
here are some example cases that both the inputs and outputs should handle: https://www.internalfb.com/intern/anp/view/?id=1672627

vkuzo commented:
here are some example cases that both the inputs and outputs should handle: https://www.internalfb.com/intern/anp/view/?id=1672627

vkuzo commented:
> 'input_quantized_idxs': {0: torch.quint8}, 'output_quantized_idxs': {0: torch.quint8}


I'm not sure that the current solution in this diff can handle a lot of the use cases in my previous comment, seems like this is hardcoding the input to a dictionary.  Should we consider a format which is more general?

vkuzo commented:
> 'input_quantized_idxs': {0: torch.quint8}, 'output_quantized_idxs': {0: torch.quint8}


I'm not sure that the current solution in this diff can handle a lot of the use cases in my previous comment, seems like this is hardcoding the input to a dictionary.  Should we consider a format which is more general?

terrychenism committed:
Update on "[quant][fx] Enable input and output idx support dtype"


Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34660121](https://our.internmc.facebook.com/intern/diff/D34660121)

[ghstack-poisoned]

terrychenism commented:
> > 'input_quantized_idxs': {0: torch.quint8}, 'output_quantized_idxs': {0: torch.quint8}
> 
> I'm not sure that the current solution in this diff can handle a lot of the use cases in my previous comment, seems like this is hardcoding the input to a dictionary. Should we consider a format which is more general?

it cannot support the last case of your notebook. No.1-4 example should work

terrychenism committed:
Update on "[quant][fx] Enable input and output idx support dtype"


Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34660121](https://our.internmc.facebook.com/intern/diff/D34660121)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34660121).

terrychenism committed:
Update on "[quant][fx] Enable input and output idx support dtype"


Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34660121](https://our.internmc.facebook.com/intern/diff/D34660121)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34660121).

terrychenism committed:
Update on "[quant][fx] Enable input and output idx support dtype"


Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34660121](https://our.internmc.facebook.com/intern/diff/D34660121)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34660121).

vkuzo commented:
> it cannot support the last case of your notebook

Doesn't the current format only support lists without nesting?  Could you share some examples on how this would handle other formats?

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -774,10 +766,23 @@ def _recursive_maybe_replace_node_with_obs(
+                if arg_idx in output_quantized_idxs:
+                    output_target_dtype = output_quantized_idxs[arg_idx]
+                else:
+                    output_target_dtype = torch.quint8
```
shouldn't this be `torch.float`?

terrychenism replied:
i think it should be quint8, see   L711 the original default output_target_dtype = torch.quint8

jerryzh168 replied:
L711 is set when we have output_quantized_idxs == [0] I think, in this branch the scenario is that there are multiple outputs, let's say 2 output, but user only specifies the dtype for one of them, e.g. {0: torch.quint8}, then for the one that user didn't mention, we should set the dtype to float I think. You should have a test case for this as well to catch this case I think

terrychenism replied:
I mean L711 in original code, it used quint8 by default.

jerryzh168 replied:
yeah I'm talking about original code as well, given the assert in L703 and return in L708, we are sure at that point `output_quantized_idxs == [0]`

terrychenism replied:
Got it. 
Dict support is already, check the code from L779

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -774,10 +766,23 @@ def _recursive_maybe_replace_node_with_obs(
-            _recursive_maybe_replace_node_with_obs(
-                old_arg, output_target_dtype, node_name_to_target_dtype,
-                qconfig_map, model, modules, graph))
+        if isinstance(old_arg, (list, tuple)):
```
what if we return a `dict`, what does it look like?

terrychenism replied:
do we need to support dict as well? 
list and tuple are common outputs, i didn't see the use case for output the dict of tensor.

jerryzh168 replied:
even if we don't support it, I feel we should test and maybe add an assert to be clear

jerryzh168 replied:
would be good to add a dict output test case as well

terrychenism replied:
Sure will add it

jerryzh168 replied:
btw, we do have dict output use cases in d2go, I'm adding a fix in convert right now

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -1421,6 +1426,6 @@ def prepare(
         # Union[Tensor, Module]
         model._standalone_module_input_quantized_idxs = \
-            torch.tensor(input_quantized_idxs)
-        model._standalone_module_output_quantized_idxs = torch.tensor(output_quantized_idxs)
+            torch.tensor(list(input_quantized_idxs.keys()))
+        model._standalone_module_output_quantized_idxs = torch.tensor(list(output_quantized_idxs.keys()))
```
why are we using keys here? should we use dict here as well?

terrychenism replied:
following the original idx, we should use 'keys' here.

jerryzh168 replied:
I feel we should keep it as dict and change the code that is handling this

jerryzh168 replied:
we can do this after my PR is landed: https://github.com/pytorch/pytorch/pull/73863, currently it is expecting a list of indicies I think.

jerryzh168 replied:
we can do this after my PR is landed: https://github.com/pytorch/pytorch/pull/73863, currently it is expecting a list of indicies I think.

terrychenism replied:
sounds good. I'll keep 'keys' in this PR.

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2355,6 +2355,124 @@ def test_fp32_input_fp32_output(self):
+            'output_quantized_idxs': {0: torch.quint8, 1: torch.quint8}
+        }
+        prepare_count_check = {
+            ns.call_module(torch.ao.quantization.MinMaxObserver): 2,
```
can we add more comments for these expected results, e.g. where do we expect to see observers being inserted? it would make the test case much easier to understand

jerryzh168 replied:
can we add more comments for these expected results, e.g. where do we expect to see observers being inserted? it would make the test case much easier to understand

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2355,6 +2355,124 @@ def test_fp32_input_fp32_output(self):
+            ns.call_module(torch.ao.quantization.MinMaxObserver): 2,
+        }
+        convert_count_check = {
+            ns.call_function(torch.quantize_per_tensor): 0,
```
also can we add some comments about why these should be 0 as well?

jerryzh168 replied:
also can we add some comments about why these should be 0 as well?

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2355,6 +2355,124 @@ def test_fp32_input_fp32_output(self):
+            'output_quantized_idxs': {1: torch.quint8}
+        }
+        prepare_count_check = {
+            ns.call_module(torch.ao.quantization.MinMaxObserver): 2,
```
should this be `1` or `2`?

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2306,7 +2306,7 @@ def forward(self, x):
 
     def test_quantized_input_quantized_output(self):
```
I think we can also add a test case similar to `test_quantized_input_quantized_multiple_output` for standalone modules

jerryzh168 replied:
`test_standalone_module_quantized_interface` these tests are trying to test the standalone module api. maybe I can explain in vc

jerryzh168 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -2306,7 +2306,7 @@ def forward(self, x):
 
     def test_quantized_input_quantized_output(self):
```
I think we can also add a test case similar to `test_quantized_input_quantized_multiple_output` for standalone modules

jerryzh168 reviewed (changes requested):
still require some changes to land I think, had some inline comments

terrychenism committed:
Update on "[quant][fx] Enable input and output idx support dtype"


Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34660121](https://our.internmc.facebook.com/intern/diff/D34660121)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][fx] Enable input and output idx support dtype"


Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34660121](https://our.internmc.facebook.com/intern/diff/D34660121)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][fx] Enable input and output idx support dtype"


Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34660121](https://our.internmc.facebook.com/intern/diff/D34660121)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34660121).

terrychenism committed:
Update on "[quant][fx] Enable input and output idx support dtype"


Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34660121](https://our.internmc.facebook.com/intern/diff/D34660121)

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/prepare.py:
```diff
@@ -774,10 +766,34 @@ def _recursive_maybe_replace_node_with_obs(
+                        qconfig_map, model, modules, graph))
+        elif isinstance(old_arg, dict):
+            results_dict = {}
+            for arg_idx, inner_v in enumerate(old_arg.items()):
```
do we want to support {"key": dtype} or just treat the whole dictionary as one unit?

terrychenism committed:
Update on "[quant][fx] Enable input and output idx support dtype"


Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34660121](https://our.internmc.facebook.com/intern/diff/D34660121)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][fx] Enable input and output idx support dtype"


Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34660121](https://our.internmc.facebook.com/intern/diff/D34660121)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34660121).

terrychenism committed:
Update on "[quant][fx] Enable input and output idx support dtype"


Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34660121](https://our.internmc.facebook.com/intern/diff/D34660121)

[ghstack-poisoned]

terrychenism committed:
Update on "[quant][fx] Enable input and output idx support dtype"


Summary:
Enable input and output idx support dtype

Test Plan:
python3 test/test_quantization.py TestQuantizeFx.test_quantized_input_quantized_multiple_output

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34660121](https://our.internmc.facebook.com/intern/diff/D34660121)

[ghstack-poisoned]

terrychenism commented:
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34660121).
